wallets/dcr/rescan.go

```diff
@@ -0,0 +1,141 @@
+package dcr
+
+import (
+	"context"
+	"math"
+	"time"
+
+	"decred.org/dcrwallet/v2/errors"
+	w "decred.org/dcrwallet/v2/wallet"
+)
+
+func (wallet *Wallet) RescanBlocks(walletID int) error {
+	return wallet.RescanBlocksFromHeight(walletID, 0)
+}
+
+func (wallet *Wallet) RescanBlocksFromHeight(walletID int, startHeight int32) error {
+
+	netBackend, err := wallet.Internal().NetworkBackend()
+	if err != nil {
+		return errors.E(ErrNotConnected)
+	}
+
+	if wallet.IsRescanning() || !wallet.IsSynced() {
+		return errors.E(ErrInvalid)
+	}
+
+	go func() {
+		defer func() {
+			wallet.syncData.mu.Lock()
+			wallet.syncData.rescanning = false
+			wallet.syncData.cancelRescan = nil
+			wallet.syncData.mu.Unlock()
+		}()
+
+		ctx, cancel := wallet.ShutdownContextWithCancel()
+
+		wallet.syncData.mu.Lock()
+		wallet.syncData.rescanning = true
+		wallet.syncData.cancelRescan = cancel
+		wallet.syncData.mu.Unlock()
+
+		if wallet.blocksRescanProgressListener != nil {
+			wallet.blocksRescanProgressListener.OnBlocksRescanStarted(walletID)
+		}
+
+		progress := make(chan w.RescanProgress, 1)
+		go wallet.Internal().RescanProgressFromHeight(ctx, netBackend, startHeight, progress)
+
+		rescanStartTime := time.Now().Unix()
+
+		for p := range progress {
+			if p.Err != nil {
+				log.Error(p.Err)
+				if wallet.blocksRescanProgressListener != nil {
+					wallet.blocksRescanProgressListener.OnBlocksRescanEnded(walletID, p.Err)
+				}
+				return
+			}
+
+			rescanProgressReport := &HeadersRescanProgressReport{
+				CurrentRescanHeight: p.ScannedThrough,
+				TotalHeadersToScan:  wallet.getBestBlock(),
+				WalletID:            walletID,
+			}
+
+			elapsedRescanTime := time.Now().Unix() - rescanStartTime
+			rescanRate := float64(p.ScannedThrough) / float64(rescanProgressReport.TotalHeadersToScan)
+
+			rescanProgressReport.RescanProgress = int32(math.Round(rescanRate * 100))
+			estimatedTotalRescanTime := int64(math.Round(float64(elapsedRescanTime) / rescanRate))
+			rescanProgressReport.RescanTimeRemaining = estimatedTotalRescanTime - elapsedRescanTime
+
+			rescanProgressReport.GeneralSyncProgress = &GeneralSyncProgress{
+				TotalSyncProgress:         rescanProgressReport.RescanProgress,
+				TotalTimeRemainingSeconds: rescanProgressReport.RescanTimeRemaining,
+			}
+
+			if wallet.blocksRescanProgressListener != nil {
+				wallet.blocksRescanProgressListener.OnBlocksRescanProgress(rescanProgressReport)
```

Comment:
IMO dropping the word `rescan` from variable names would make the variables shorter and still retain a descriptive meaning.

Reply:
as of now, i'm not writing any new code, i just moved all the old code into a package, perhaps we can consider this during a code cleanup